src/main/proto/com/google/api/codegen/config.proto

```diff
@@ -316,62 +318,77 @@ message CodeSampleValues {
-    // `MethodConfigProto.code_sample_values.id`. If any element of
-    // `ids` is invalid (i.e., is not found in
-    // `MethodConfigProto.code_sample_values.id`) and `delete` is not
-    // set, then an fatal error is raised at code-generation time.
-    repeated string ids = 1;
-
-    // Whether the samples specified by this configuration should also
-    // be used as configurable "Explorer Samples" suitable for
-    // inclusion in a UI element.
-    bool explorer_sample = 2;
-
-    // Whether the samples specified by this configuration should also
-    // be included within code code comments in the surface of each
-    // generated client library.
-    bool in_code_sample = 3;
-
-    // A hint as to the user-visible position of this
-    // `SampleConfiguration` relative to other
-    // `SampleConfiguration`s. This is only used when generating
-    // "Explorer Samples", which only happens if `explorer_sample` is
-    // set.
-    uint32 position = 4;
-  }
 
-  // The sample code mold (defined in the client library generator for
-  // various method types) to which this `SampleConfiguration` applies
-  //
-  // TODO(vchudnov-g): Point to a reference listing the possible molds
-  // for each method type.
-  repeated string molds = 1;
+  // `SampleTypeConfiguration` defines the samples that will show up
+  // as a particular sample type. If order matters for this sample
+  // type, the order of the samples is taken to be primarily the order
+  // in which the `calling_forms` appear, and secondarily, for each
+  // such form, the order in which the `value_sets` appear.
+  message SampleTypeConfiguration {
+    // One or more expressions matching the method calling forms
+    // (defined by the client generator) applicable to this
+    // method. Any expressions that do not match at least one calling
+    // form cause a fatal error during code generation. If not
+    // specified, samples are generated for all such forms.
+    //
+    // An expression matches an ID:
+    //  (a) if the expression equals "*" (which matches all IDs), or
+    //  (b) if the expression equals the ID identically, or
+    //  (c) if the ID equals the expression concatenated with a "#" character
+    //      followed by any number of other characters.
+    //
+    // TODO(vchudnov-g): Point to a reference listing the possible calling forms
+    // for each method type.
+    repeated string calling_forms = 1;
+
+    // One or more expressions matching the
+    // `MethodConfigProto.sample_value_sets.id`s defined for this
+    // method. Any expressions on this list that do not match a
+    // `SampleValueSet.id` cause a fatal error during code
+    // generation. If not specified or if empty, no samples of this
+    // type will be generated.
+    //
+    // An expression matches an ID
+    // (a) if the expression equals "*" (which matches all IDs), or
+    // (b) if the leading characters of the ID are identical to the
```

Comment:
When do you anticipate this "prefix matching" feature being used?

Did you consider making `*` a wildcard in the expression to make these semantics more explicit (i.e., expression `foo*` matches ID `foobar`, but expression `foo` does not match ID `foobar`)?

Reply:
I did, but I was trying to hew to the mental model in the document, where we use `#` separated fields to specify the method style and the calling form. However, what we ultimately care about in the config is the method form, so maybe it's easier to grok simply having wildcard matching here. The organization of the styles and forms can be completely contained in our naming convention for them, which we control.....

OK, changing it.